bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,820 @@
+    ErrorResponseReceived {
+        from: Baseboard,
+        state: &'static str,
+        request_id: Uuid,
+        error: MsgError,
+    },
+}
+
+pub struct Fsm {
+    /// The current state of this peer
+    state: State,
+    /// Unique ID of this peer
+    id: Baseboard,
+
+    /// User provided configuration
+    config: Config,
+
+    /// Unique IDs of connected peers
+    connected_peers: BTreeSet<Baseboard>,
+
+    /// Manage all trackable requests
+    request_manager: RequestManager,
+
+    /// Envelopes not managed by the `RequestManager`
+    /// These are all envelopes containing `Response` messages
+    envelopes: Vec<Envelope>,
+
+    /// We keep track of whether the rack failed to initialize. If this happens
+    /// the coordinator should return this error on every new API request.
+    rack_init_error: Option<(Uuid, ApiError)>,
+}
+
+impl Fsm {
+    /// Create a new FSM in `State::Uninitialized`
+    pub fn new_uninitialized(id: Baseboard, config: Config) -> Fsm {
+        Fsm {
+            state: State::Uninitialized,
+            id: id.clone(),
+            config,
+            connected_peers: BTreeSet::new(),
+            request_manager: RequestManager::new(id, config),
+            envelopes: vec![],
+            rack_init_error: None,
+        }
+    }
+
+    pub fn config(&self) -> &Config {
+        &self.config
+    }
+
+    /// Return any envelopes that need sending
+    ///
+    /// This must be called after any API callback
+    pub fn drain_envelopes(&mut self) -> impl Iterator<Item = Envelope> + '_ {
+        self.envelopes.drain(..).chain(self.request_manager.drain_elements())
+    }
+
+    /// Put a request into an envelope and add it to `self.envelopes`
+    pub fn push_response(
+        &mut self,
+        to: Baseboard,
+        request_id: Uuid,
+        type_: ResponseType,
+    ) {
+        self.envelopes
+            .push(Envelope { to, msg: Response { request_id, type_ }.into() });
+    }
+
+    pub fn state(&self) -> &State {
+        &self.state
+    }
+
+    pub fn rack_init_failed(&self) -> bool {
+        self.rack_init_error.is_some()
+    }
+
+    /// This call is triggered locally on a single sled as a result of RSS
+    /// running. It may only be called once, which is enforced by checking to
+    /// see if we already are in `State::Uninitialized`.
+    ///
+    /// Persistence is required after a successful call to `init_rack`.
+    pub fn init_rack(
+        &mut self,
+        now: Instant,
+        rack_uuid: Uuid,
+        initial_membership: BTreeSet<Baseboard>,
+    ) -> Result<(), ApiError> {
+        if let Some((_, err)) = &self.rack_init_error {
+            return Err(err.clone());
+        }
+        let State::Uninitialized = self.state else {
+            return Err(ApiError::AlreadyInitialized);
+        };
+        let pkgs = create_pkgs(rack_uuid, initial_membership.clone())
+            .map_err(|e| ApiError::RackInitFailed(e))?;
```

Comment:
Nit: I think you could do
```suggestion
            .map_err(ApiError::RackInitFailed)?;
```